File: ZipTools.vb
Imports System.IO
Imports PRISM.Processes
Imports PRISM.Logging

Namespace Files
	''' <summary>Makes using a file archiving program easier.</summary>
	''' <remarks>There are a routines to create an archive, extract files from an existing archive,
    ''' and to verify an existing archive.
	''' </remarks>
	Public Class ZipTools
		Implements ILoggerAware

		Private m_WorkDir As String
		Private m_ZipFilePath As String
        Private ReadOnly m_WaitInterval As Integer
		Private m_ExceptionLogger As ILogger
        Private m_EventLogger As ILogger
        Private m_CreateNoWindow As Boolean
        Private m_WindowStyle As ProcessWindowStyle

        ''' <summary>Create a zip file.</summary>
        ''' <param name="CmdOptions">The zip program command line arguments.</param>
        ''' <param name="OutputFile">The file path of the output zip file.</param>
        ''' <param name="InputSpec">The files and/or directorys to archive.</param>
        Public Function MakeZipFile(
          CmdOptions As String,
          OutputFile As String,
          InputSpec As String) As Boolean


            'Verify input file and output path have been specified
            If (m_ZipFilePath = "") Or (m_WorkDir = "") Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Input file path and/or working path not specified.", ILogger.logMsgType.logError, True)
                End If
                Return False
            End If

            'Setup the zip program
            Dim zipper = New clsProgRunner() With {
                .Arguments = "-Add " & CmdOptions & " """ & OutputFile & """ """ & InputSpec & """",
                .Program = m_ZipFilePath,
                .WorkDir = m_WorkDir,
                .MonitoringInterval = m_WaitInterval,
                .Name = "Zipper",
                .Repeat = False,
                .RepeatHoldOffTime = 0,
                .CreateNoWindow = m_CreateNoWindow
            }

            'Start the zip program
            zipper.StartAndMonitorProgram()

            'Wait for zipper program to complete
            While (zipper.State <> clsProgRunner.States.NotMonitoring)
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Waiting for zipper program.  Going to sleep for " & m_WaitInterval & " milliseconds.",
                      ILogger.logMsgType.logHealth, True)
                End If
                Threading.Thread.Sleep(m_WaitInterval)
            End While

            'Check for valid return value after completion
            If zipper.ExitCode <> 0 Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Zipper program exited with code: " & zipper.ExitCode, ILogger.logMsgType.logError, True)
                End If
                Return False
            Else
                Return True
            End If

        End Function

        ''' <summary>Extract files from a zip file.</summary>
        ''' <param name="CmdOptions">The zip program command line arguments.</param>
        ''' <param name="InputFile">The file path of the zip file from which to extract files.</param>
        ''' <param name="OutPath">The path where you want to put the extracted files.</param>
        Public Function UnzipFile(CmdOptions As String, InputFile As String, OutPath As String) As Boolean


            'Verify input file and output path have been specified
            If (m_ZipFilePath = "") Or (m_WorkDir = "") Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Input file path and/or working path not specified.", ILogger.logMsgType.logError, True)
                End If
                Return False
            End If

            'Verify input file exists
            If Not File.Exists(InputFile) Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Input file " & InputFile & " not found", ILogger.logMsgType.logError, True)
                End If
                Return False
            End If

            'Verify output path exists
            If Not Directory.Exists(OutPath) Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Output directory " & OutPath & " does not exist.", ILogger.logMsgType.logError, True)
                End If
                Return False
            End If

            'Setup the unzip program
            Dim zipper = New clsProgRunner() With {
                .Arguments = "-Extract " & CmdOptions & " """ & InputFile & """ """ & OutPath & """",
                .MonitoringInterval = m_WaitInterval,
                .Name = "Zipper",
                .Program = m_ZipFilePath,
                .WorkDir = m_WorkDir,
                .Repeat = False,
                .RepeatHoldOffTime = 0,
                .CreateNoWindow = m_CreateNoWindow,
                .WindowStyle = m_WindowStyle
            }

            'Start the unzip program
            zipper.StartAndMonitorProgram()

            'Wait for zipper program to complete
            While (zipper.State <> clsProgRunner.States.NotMonitoring)
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Waiting for zipper program.  Going to sleep for " & m_WaitInterval & " milliseconds.",
                       ILogger.logMsgType.logHealth, True)
                End If
                Threading.Thread.Sleep(m_WaitInterval)
            End While

            'Check for valid return value after completion
            If zipper.ExitCode <> 0 Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Zipper program exited with code: " & zipper.ExitCode, ILogger.logMsgType.logError, True)
                End If
                Return False
            Else
                Return True
            End If

        End Function

        ''' <summary>Defines whether a window is displayed when calling the zipping program.</summary>
        Public Property CreateNoWindow() As Boolean
            Get
                Return m_CreateNoWindow
            End Get
            Set(Value As Boolean)
                m_CreateNoWindow = Value
            End Set
        End Property

        ''' <summary>
        ''' Window style to use when CreateNoWindow is False.
        ''' </summary>
        Public Property WindowStyle() As ProcessWindowStyle
            Get
                Return m_WindowStyle
            End Get
            Set(Value As ProcessWindowStyle)
                m_WindowStyle = Value
            End Set
        End Property


        ''' <summary>The working directory for the zipping process.</summary>
        Public Property WorkDir() As String
            Get
                Return m_WorkDir
            End Get
            Set(Value As String)
                m_WorkDir = Value
            End Set
        End Property

        ''' <summary>The path to the zipping program.</summary>
        Public Property ZipFilePath() As String
            Get
                Return m_ZipFilePath
            End Get
            Set(Value As String)
                m_ZipFilePath = Value
            End Set
        End Property

        ''' <summary>Initializes a new instance of the ZipTools class.</summary>
        ''' <param name="WorkDir">The working directory for the zipping process.</param>
        ''' <param name="ZipFilePath">The path to the zipping program.</param>
        Public Sub New(WorkDir As String, ZipFilePath As String)
            m_WorkDir = WorkDir
            m_ZipFilePath = ZipFilePath
            m_WaitInterval = 2000  'msec
            NotifyOnEvent = True
            NotifyOnException = True
        End Sub

        ''' <summary>Verifies initial parameters have been set prior to performing operation.</summary>
        Private Function VerifyParams() As Boolean

            If (m_WorkDir = "") Or (m_ZipFilePath) = "" Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Working directory and/or zipfile's path not specified.", ILogger.logMsgType.logError, True)
                End If
                Return False
            Else
                Return True
            End If

        End Function

        ''' <summary>Verifies the integrity of a zip file.</summary>
        ''' <param name="FilePath">The file path of the zip file to verify.</param>
        Public Function VerifyZippedFile(FilePath As String) As Boolean


            'Verify test file exists
            If Not File.Exists(FilePath) Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("File path file " & FilePath & " not found", ILogger.logMsgType.logError, True)
                End If
                Return False
            End If

            'Verify Zip file and output path have been specified
            If (m_ZipFilePath = "") Or (m_WorkDir = "") Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Zip file path and/or working path not specified.", ILogger.logMsgType.logError, True)
                End If
                Return False
            End If

            'Setup the zip program
            Dim zipper = New clsProgRunner() With {
                .Arguments = "-test -nofix" & " " & FilePath,
                .Program = m_ZipFilePath,
                .WorkDir = m_WorkDir,
                .MonitoringInterval = m_WaitInterval,
                .Name = "Zipper",
                .Repeat = False,
                .RepeatHoldOffTime = 0,
                .CreateNoWindow = m_CreateNoWindow,
                .WindowStyle = m_WindowStyle
            }

            'Start the zip program
            zipper.StartAndMonitorProgram()

            'Wait for zipper program to complete
            While (zipper.State <> clsProgRunner.States.NotMonitoring)
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Waiting for zipper program.  Going to sleep for " & m_WaitInterval & " milliseconds.",
                       ILogger.logMsgType.logHealth, True)
                End If
                Threading.Thread.Sleep(m_WaitInterval)
            End While

            'Check for valid return value after completion
            If zipper.ExitCode <> 0 Then
                If Not m_EventLogger Is Nothing Then
                    m_EventLogger.PostEntry("Zipper program exited with code: " & zipper.ExitCode, ILogger.logMsgType.logError, True)
                End If
                Return False
            Else
                Return True
            End If

        End Function

        ''' <summary>Sets the name of the exception logger</summary>
        Public Sub RegisterExceptionLogger(logger As ILogger) Implements ILoggerAware.RegisterEventLogger
            m_ExceptionLogger = logger
        End Sub

        ''' <summary>Sets the name of the event logger</summary>
        Public Sub RegisterEventLogger(logger As ILogger) Implements ILoggerAware.RegisterExceptionLogger
            m_EventLogger = logger
        End Sub

        ''' <summary>Gets or Sets notify on event.</summary>
        Public Property NotifyOnEvent As Boolean Implements ILoggerAware.NotifyOnEvent

        ''' <summary>Gets or Sets notify on exception.</summary>
        Public Property NotifyOnException As Boolean Implements ILoggerAware.NotifyOnException

    End Class
End Namespace

File: ZipTools.cs
using System;
using System.Diagnostics;
using System.IO;
using PRISM.Logging;

namespace PRISM
{
    /// <summary>
    /// Makes using a file archiving program easier.
    /// </summary>
    /// <remarks>There are a routines to create an archive, extract files from an existing archive,
    /// and to verify an existing archive.
    /// </remarks>
    public class ZipTools
    {
        /// <summary>
        /// Working directory
        /// </summary>
        private string m_WorkDir;

        /// <summary>
        /// Program used for zipping and unzipping files
        /// </summary>
        private string m_ZipProgramPath;

        /// <summary>
        /// Interval, in milliseconds, to sleep between checking the status of a zip or unzip task
        /// </summary>
        private readonly int m_WaitInterval;

        /// <summary>
        /// Logging class
        /// </summary>
        private BaseLogger m_Logger;

#pragma warning disable 618
        [Obsolete("Use m_Logger (typically a FileLogger)")]
        private ILogger m_EventLogger;
#pragma warning restore 618

        private bool m_CreateNoWindow;

        private ProcessWindowStyle m_WindowStyle;

        /// <summary>
        /// Create a zip file.
        /// </summary>
        /// <param name="cmdOptions">The zip program command line arguments.</param>
        /// <param name="outputFile">The file path of the output zip file.</param>
        /// <param name="inputSpec">The files and/or directories to archive.</param>
        public bool MakeZipFile(string cmdOptions, string outputFile, string inputSpec)
        {


            // Verify input file and output path have been specified
            if (string.IsNullOrEmpty(m_ZipProgramPath) || string.IsNullOrEmpty(m_WorkDir))
            {
                var msg = "Zip program path and/or working path not specified";
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logError, true);
#pragma warning restore 618
                m_Logger?.Error(msg);

                return false;
            }

            // Setup the zip program
            var zipper = new clsProgRunner
            {
                Arguments = "-Add " + cmdOptions + " \"" + outputFile + "\" \"" + inputSpec + "\"",
                Program = m_ZipProgramPath,
                WorkDir = m_WorkDir,
                MonitoringInterval = m_WaitInterval,
                Name = "Zipper",
                Repeat = false,
                RepeatHoldOffTime = 0,
                CreateNoWindow = m_CreateNoWindow
            };

            // Start the zip program
            zipper.StartAndMonitorProgram();

            // Wait for zipper program to complete
            var success = WaitForZipProgram(zipper);

            return success;
        }

        /// <summary>
        /// Extract files from a zip file.
        /// </summary>
        /// <param name="cmdOptions">The zip program command line arguments.</param>
        /// <param name="zipFilePath">The file path of the zip file from which to extract files.</param>
        /// <param name="outFolderPath">The path where you want to put the extracted files.</param>
        public bool UnzipFile(string cmdOptions, string zipFilePath, string outFolderPath)
        {

            // Verify input file and output path have been specified
            if (string.IsNullOrEmpty(m_ZipProgramPath) || string.IsNullOrEmpty(m_WorkDir))
            {
                var msg = "Zip program path and/or working path not specified";
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logError, true);
#pragma warning restore 618
                m_Logger?.Error(msg);

                return false;
            }

            // Verify input file exists
            if (!File.Exists(zipFilePath))
            {
                var msg = "Input file not found: " + zipFilePath;
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logError, true);
#pragma warning restore 618
                m_Logger?.Error(msg);

                return false;
            }

            // Verify output path exists
            if (!Directory.Exists(outFolderPath))
            {
                var msg = "Output directory " + outFolderPath + " does not exist";
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logError, true);
#pragma warning restore 618
                m_Logger?.Error(msg);

                return false;
            }

            // Setup the unzip program
            var zipper = new clsProgRunner
            {
                Arguments = "-Extract " + cmdOptions + " \"" + zipFilePath + "\" \"" + outFolderPath + "\"",
                MonitoringInterval = m_WaitInterval,
                Name = "Zipper",
                Program = m_ZipProgramPath,
                WorkDir = m_WorkDir,
                Repeat = false,
                RepeatHoldOffTime = 0,
                CreateNoWindow = m_CreateNoWindow,
                WindowStyle = m_WindowStyle,
            };

            // Start the unzip program
            zipper.StartAndMonitorProgram();

            // Wait for zipper program to complete
            var success = WaitForZipProgram(zipper);

            return success;
        }

        /// <summary>
        /// Defines whether a window is displayed when calling the zipping program.
        /// </summary>
        public bool CreateNoWindow
        {
            get => m_CreateNoWindow;
            set => m_CreateNoWindow = value;
        }

        /// <summary>
        /// Window style to use when CreateNoWindow is False.
        /// </summary>
        public ProcessWindowStyle WindowStyle
        {
            get => m_WindowStyle;
            set => m_WindowStyle = value;
        }

        /// <summary>
        /// The working directory for the zipping process.
        /// </summary>
        public string WorkDir
        {
            get => m_WorkDir;
            set => m_WorkDir = value;
        }

        /// <summary>
        /// The path to the zipping program.
        /// </summary>
        public string ZipFilePath
        {
            get => m_ZipProgramPath;
            set => m_ZipProgramPath = value;
        }

        /// <summary>
        /// Initializes a new instance of the ZipTools class.
        /// </summary>
        /// <param name="workDir">The working directory for the zipping process.</param>
        /// <param name="zipFilePath">The path to the zipping program.</param>
        public ZipTools(string workDir, string zipFilePath)
        {
            m_WorkDir = workDir;
            m_ZipProgramPath = zipFilePath;

            // Time in milliseconds
            m_WaitInterval = 2000;

            NotifyOnEvent = true;
            NotifyOnException = true;
        }

        /// <summary>
        /// Verifies the integrity of a zip file.
        /// </summary>
        /// <param name="zipFilePath">The file path of the zip file to verify.</param>
        public bool VerifyZippedFile(string zipFilePath)
        {

            // Verify test file exists
            if (!File.Exists(zipFilePath))
            {
                var msg = "Zip file not found; cannot verify: " + zipFilePath;
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logError, true);
#pragma warning restore 618
                m_Logger?.Error(msg);

                return false;
            }

            // Verify Zip file and output path have been specified
            if (string.IsNullOrEmpty(m_ZipProgramPath) || string.IsNullOrEmpty(m_WorkDir))
            {
                var msg = "Zip program path and/or working path not specified";
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logError, true);
#pragma warning restore 618
                m_Logger?.Error(msg);

                return false;
            }

            // Setup the zip program
            var zipper = new clsProgRunner
            {
                Arguments = "-test -nofix" + " " + zipFilePath,
                Program = m_ZipProgramPath,
                WorkDir = m_WorkDir,
                MonitoringInterval = m_WaitInterval,
                Name = "Zipper",
                Repeat = false,
                RepeatHoldOffTime = 0,
                CreateNoWindow = m_CreateNoWindow,
                WindowStyle = m_WindowStyle,
            };

            // Start the zip program
            zipper.StartAndMonitorProgram();

            // Wait for zipper program to complete
            var success = WaitForZipProgram(zipper);

            return success;
        }

        private bool WaitForZipProgram(clsProgRunner zipper)
        {
            while (zipper.State != clsProgRunner.States.NotMonitoring)
            {
                var msg = "Waiting for zipper program; sleeping for " + m_WaitInterval + " milliseconds";
#pragma warning disable 618
                m_EventLogger?.PostEntry(msg, logMsgType.logHealth, true);
#pragma warning restore 618
                m_Logger?.Debug(msg);

                clsProgRunner.SleepMilliseconds(m_WaitInterval);
            }

            // Check for valid return value after completion
            if (zipper.ExitCode == 0)
                return true;

            var errorMsg = "Zipper program exited with code: " + zipper.ExitCode;
#pragma warning disable 618
            m_EventLogger?.PostEntry(errorMsg, logMsgType.logError, true);
#pragma warning restore 618
            m_Logger?.Error(errorMsg);

            return false;
        }

        /// <summary>
        /// Associate a logger with this class
        /// </summary>
        public void RegisterEventLogger(BaseLogger logger)
        {
            m_Logger = logger;
        }

        /// <summary>
        /// Associate an event logger with this class
        /// </summary>
        [Obsolete("Use RegisterEventLogger that takes a BaseLogger (typically a FileLogger)")]
        public void RegisterEventLogger(ILogger logger)
        {
            m_EventLogger = logger;
        }

        /// <summary>
        /// Gets or Sets notify on event.
        /// </summary>
        public bool NotifyOnEvent { get; set; }

        /// <summary>
        /// Gets or Sets notify on exception.
        /// </summary>
        public bool NotifyOnException { get; set; }

    }
}
